File: web/src/instruction/taskExecute.ts
import { Program } from "@project-serum/anchor";
import {
  PublicKey,
  SYSVAR_CLOCK_PUBKEY,
  TransactionInstruction,
} from "@solana/web3.js";
import { buildRemainingAccounts } from "./utils";
import { Cronos } from "../idl";
import { Account } from "../account";

export type TaskExecuteArgs = {
  task: PublicKey;
  worker: PublicKey;
};

export class TaskExecute {
  private account: Account;
  private cronos: Program<Cronos>;

  constructor(account: Account, cronos: Program<Cronos>) {
    this.account = account;
    this.cronos = cronos;
  }

  public async taskExecute({
    task,
    worker,
  }: TaskExecuteArgs): Promise<TransactionInstruction> {
    const configPDA = await this.account.config.pda();
    const taskData = await this.account.task.data(task);
    const feePDA = await this.account.fee.pda(taskData.daemon);
    return this.cronos.instruction.taskExecute({
      accounts: {
        clock: SYSVAR_CLOCK_PUBKEY,
        config: configPDA.address,
        daemon: taskData.daemon,
        fee: feePDA.address,
        task: task,
        worker: worker,
      },
      remainingAccounts: buildRemainingAccounts(taskData.ix, taskData.daemon),
    });
  }
}

File: sdk/src/instruction/task_execute.rs
use anchor_client::anchor_lang::{
    solana_program::{
        instruction::{AccountMeta, Instruction},
        pubkey::Pubkey,
        sysvar,
    },
    InstructionData,
};

pub fn task_execute(
    config: Pubkey,
    daemon: Pubkey,
    fee: Pubkey,
    task: Pubkey,
    worker: Pubkey,
) -> Instruction {
    Instruction {
        program_id: cronos_program::ID,
        accounts: vec![
            AccountMeta::new_readonly(sysvar::clock::ID, false),
            AccountMeta::new_readonly(config, false),
            AccountMeta::new(daemon, false),
            AccountMeta::new(fee, false),
            AccountMeta::new(task, false),
            AccountMeta::new(worker, true),
        ],
        data: cronos_program::instruction::TaskExecute {}.data(),
    }
}
